app/src/main/java/com/simonramstedt/yoke/YokeActivity.java

```diff
@@ -142,19 +157,37 @@ private void deleteRecursively(File joypadPath) throws IOException {
         }
 
         // Webpage uses this method to update joypad state:
+        // TODO: Investigate a WebMessagePort-based solution which works for Lollipop and older
         @JavascriptInterface
         public void update_vals(String vals) {
-            vals_str = vals;
+            vals_buffer = vals.getBytes(Charset.forName("ISO-8859-1"));
```

Comment:
Why is `Charset.forName("ISO-8859-1")` preferred over `StandardCharsets.UTF_8` or `StandardCharsets.US_ASCII`?

Reply:
Both here and in the JS side, we need a Charset that assigns a character to every byte (0x00 to 0xFF) and viceversa,

UTF-8 is a variable-length encoding. Some characters map to 2-byte sequences or more, and some bytes don't map to any characters. The result is that Python receives more bytes than expected, the `struct.unpack()` will raise an exception and the whole program crashes.

US_ASCII only maps 128 characters to the first 128 bytes (0x00 to 0x7F); as soon as we tried to decode any character outside that range, either: a) the app would crash, or b) the first bit in every byte would be lost, which would tamper with the controls.

Reply:
Ah, I see. So, if I understand correctly, we are building joypad state as ISO-8859 garbage string, then converting it to UTF-16 on JS side via [TextDecoder](https://developer.mozilla.org/en-US/docs/Web/API/TextDecoder/decode), then back to bytes on Java side. That's cool.

We are breaking compatibility with Android < 5.0 with it, though, as its native webview doesn't have TextDecoder. In my personal opinion this is okay, and we should consider bumping the required Android version to 6.0 if it'll help with development.

Reply:
HAHAHAHAH. That's... almost correct, yes. The joypad state is built as pure binary data, and then converted to a garbage string. Which is even worse.

According to `build.gradle`, this was already incompatible with Android < 5.0 (probably because of a WebView problem). At this point, Lollipop is ancient history anyways, so... I'd actually be happy to bump the Android minimum version.

I'll try the MessagePort approach and push a couple commits to each repository if I make it.